source/adapters/opencl/context.cpp

```diff
@@ -90,13 +74,16 @@ urContextGetInfo(ur_context_handle_t hContext, ur_context_info_t propName,
      * queries of each device separately and building the intersection set. */
     return UR_RESULT_ERROR_INVALID_ARGUMENT;
   }
-  case UR_CONTEXT_INFO_NUM_DEVICES:
-  case UR_CONTEXT_INFO_DEVICES:
+  case UR_CONTEXT_INFO_NUM_DEVICES: {
+    return ReturnValue(hContext->DeviceCount);
+  }
+  case UR_CONTEXT_INFO_DEVICES: {
+    return ReturnValue(&hContext->Devices[0], hContext->DeviceCount);
+  }
   case UR_CONTEXT_INFO_REFERENCE_COUNT: {
     size_t CheckPropSize = 0;
-    auto ClResult =
-        clGetContextInfo(cl_adapter::cast<cl_context>(hContext), CLPropName,
-                         propSize, pPropValue, &CheckPropSize);
+    auto ClResult = clGetContextInfo(hContext->get(), CLPropName, propSize,
```

Comment:
Now we have `ur_context_handle_t` I think this should be returning its reference count not the `cl_context` reference count.

Reply:
I was planning to add the reference counting and resolve the memory leak in a separate PR following this to be more clear. but maybe it is better to extend this PR with that so will do that here.

Reply:
I'd like to avoid introducing a known memory leak to main, at any point.

Reply:
Added Ref counting to UR layer and fixed that. Checked it also with Address Sanitizer.